pkg/registry/core/service/strategy.go

```diff
@@ -226,3 +226,109 @@ func (serviceStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runt
 func (serviceStatusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
 	return validation.ValidateServiceStatusUpdate(obj.(*api.Service), old.(*api.Service))
 }
+
+// normalizeClusterIPs adjust clusterIPs based on ClusterIP
+func normalizeClusterIPs(oldSvc *api.Service, newSvc *api.Service) {
+	// In all cases here, we don't need to over-think the inputs.  Validation
+	// will be called on the new object soon enough.  All this needs to do is
+	// try to divine what user meant with these linked fields. The below
+	// is verbosely written for clarity.
+
+	if oldSvc == nil {
+		// This was a create operation.
+		// User specified singular and not plural (e.g. an old client), so init
+		// plural for them.
+		if len(newSvc.Spec.ClusterIP) > 0 && len(newSvc.Spec.ClusterIPs) == 0 {
+			newSvc.Spec.ClusterIPs = []string{newSvc.Spec.ClusterIP}
+			return
+		}
+
+		// User specified plural and not singular (new client) so init singular for them.
+		if len(newSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIP) == 0 {
+			newSvc.Spec.ClusterIP = newSvc.Spec.ClusterIPs[0]
+			return
+		}
+
+		// Either both were not specified (will be allocated) or both were
+		// specified (will be validated).
+		return
+	}
+
+	// This was an update operation.
+
+	// **** IMPORTANT *****
+	// Old clients operate only on ClusterIP. New Clients
+	// *MUST* operate on both fields (except when setting ClusterIPs for the first time)
+
+	// ClusterIPs were cleared by an old client which was trying to patch
+	// some field and didn't provide ClusterIPs
+	if len(oldSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIPs) == 0 {
+		// if ClusterIP is the same, then it is an old client trying to
+		// patch service and didn't provide ClusterIPs
+		if oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP {
+			newSvc.Spec.ClusterIPs = oldSvc.Spec.ClusterIPs
+		}
```

Comment:
> IPFamilies and IPFamily Policy does not need this.

So I mean, given a Service

    ...
    spec:
      type: ClusterIP
      clusterIP: 1.2.3.4
      clusterIPs: ["1.2.3.4"]
      ipFamilyPolicy: SingleStack
      ipFamilies: ["IPv4"]
    ...

in order to preserve compatibility with kubernetes <= 1.19, I must be able to do

    kubectl patch service foo -p '{"spec": {"type": "ExternalName", "clusterIP": ""}}'

and have that work. As I understand the code now, this would fail with "ipFamilies may not be set for ExternalName services".

Reply:
There is a hook in the same file check `clearClusterIPRelatedFields(...)` *and* there is an integration tests we have a test similar to that https://github.com/kubernetes/kubernetes/pull/91824/commits/1c865edb0fa0e9a9f03aab595edf11c76cf07808#diff-fca4029f6ac6798e271dc01013a77455R1452